### ISSUE TITLE
OOS in MP games

### ISSUE THREAD

**Issue by MrZorG33** [We-the-People-civ4col-mod/Mod#520]
I decided to create a separate issue to solve this problem. I propose to mark here all the solutions to this problem. I will also attach reports here.

today we played three. some players have very large reports (several hundred megabytes).
I apologize if there is a little confusion in the reports - I did it late at night.
[log OOS 01.zip](https://github.com/We-the-People-civ4col-mod/Mod/files/7765916/log.OOS.01.zip)
[log OOS 02.zip](https://github.com/We-the-People-civ4col-mod/Mod/files/7765920/log.OOS.02.zip)

some reports are very large even in the archive. so I am attaching a download link from google drive.
https://drive.google.com/file/d/16ZC8Lg-SqqGhSAIDHqzSHBgaXQ0OFDgl/view?usp=sharing

**Comment by raystuttgart** [comment]
@MrZorG33 
Those logs **continue writing all the time** from first MP game to last game if I am correct.'
You will have l**og entries there from years** in worst case - and even more problematic from **different release versions**.
But that does not really help us to analyze properly.

We need a **clean log** of a **single MP game** with the **newest release**.
Everything else just makes things difficult for us.

Thus ideally you guys would follow this order:

1. Delete all those log files. --> clean log files will be created
2. Start a new MP game until it desyncs.
3. Send us the clean new OOS log files

-----

**Summary:**
We need clean managable log files. Most imporantly for me is that we **just have the one single release** in there.
Not a huge pile of date from many different releases collected over years.

By the way:

- "Plains" would be best considering the log.
- "3.0.1." would however be also ok.

@Nightinggale 
Please correct me if I am wrong.

**Comment by MrZorG33** [comment]
before starting the game, the "Logs" folder was cleared. it is impossible to create an OOS report from the game menu - there is no such item.
![image](https://user-images.githubusercontent.com/61798883/147425287-7cb525e0-c960-49d4-9859-54d43045d773.png)

some of the second player's logs are very large. besides, for some reason, the first player does not have such logs. (most likely the settings were overwritten when the game was updated. when checking it, it turned out that some settings are disabled. now all are on)
[log player 1 02.zip](https://github.com/We-the-People-civ4col-mod/Mod/files/7777844/log.player.1.02.zip)
https://drive.google.com/file/d/1nFzcmueL0hpVyZkL43pekI062SNspIrI/view?usp=sharing

**Comment by MrZorG33** [comment]
another game (2 players) with OOS.
[log OOS player 0_1.zip](https://github.com/We-the-People-civ4col-mod/Mod/files/7790087/log.OOS.player.0_1.zip)
[log OOS player 0_2.zip](https://github.com/We-the-People-civ4col-mod/Mod/files/7790090/log.OOS.player.0_2.zip)
[log OOS player 1_1.zip](https://github.com/We-the-People-civ4col-mod/Mod/files/7790092/log.OOS.player.1_1.zip)
[log OOS player 1_2.zip](https://github.com/We-the-People-civ4col-mod/Mod/files/7790093/log.OOS.player.1_2.zip)

**Comment by raystuttgart** [comment]
Thanks for the logs. 👍

**Comment by MrZorG33** [comment]
after each OOS, we exit the game to clear the logs. as you requested)

**Comment by raystuttgart** [comment]
@MrZorG33 
It is not necessary. You are creating yourselfs too much effort.
I just do **not** want to have **more than 1 game** in the OOS logs.

The absolute first OOS may be most interesting though.
So if we have a separate file from the first OOS that is great. 👍 

After that all the OOS can stay in the same log though.
**Just before you start a new game**, please **delete the logs first** so they are clean.

**Comment by MrZorG33** [comment]
continuation of the game with two players
options when OOS occurs:
- when the king / natives offer something to the player
- at the beginning of the turn, one of the players is in the Europe or colony screen
[log OOS 03.zip](https://github.com/We-the-People-civ4col-mod/Mod/files/7793925/log.OOS.03.zip)
[log OOS 04.zip](https://github.com/We-the-People-civ4col-mod/Mod/files/7793922/log.OOS.04.zip)
[log OOS 05.zip](https://github.com/We-the-People-civ4col-mod/Mod/files/7793923/log.OOS.05.zip)
.

**Comment by raystuttgart** [comment]
@MrZorG33 
Thanks, I guess we now have **enough data** to analyze. 👍 
Collecting more will just be **unnecessary work** for you guys.

----

Once @Nightinggale or myself have ti**me to check** we should take a look at this data first.

**Comment by raystuttgart** [comment]
**@Nightinggale** 
Almost all of these DLL-Diplo-Events can trigger **"SorenRandNum" from the Diplomacy-Dialogue (UI)**, whenever the **Player selects a conversation choice** with King/Natives/Bishop/ ....

**e.g. all of this is randomized:**

- Number of Gold the King / Natives / Bishop want (in next choice)
- Success / Failur of Conversation Choice (of selectd option)
- Number of Units the Player will get (by selected option)
- King / Natives / Bishop being happy / angry  (by selected option)
....

Most likely we should really change all the "SorenRandNum" in those DLL-Diplomacy-Dialogues as you suggested.
(By giving **"SorenRandNum" a boolean** if it should generate the next random number or not.)

**Comment by MrZorG33** [comment]
I will clarify: OOS did not appear after some action of the player (when he chooses some option offered to him), which the king / natives offer, but immediately after the appearance of a conversation window with the king / natives.

**Comment by raystuttgart** [comment]
@MrZorG33 
Thanks, that is also very valuable information. 

@Nightinggale
Already during the calcuation of the DLL-Diplo-Events there are randoms.
Those randoms may cause the OOS issues as well.

**Comment by ErikSandbraaten** [comment]
Check out this commit in  https://github.com/We-the-People-civ4col-mod/Mod/commit/bb6c413416813e66d5d27aa539379cc015fd1cdb Rc 3.0.2 for a (partial) solution.

I've verified that the AI never calls CvPlayerAI::AI_counterPropose during autoplay so we never have to bother with a sync version.  (although the AI does call this function in Civ4).

**Comment by MrZorG33** [comment]
By the way, one more moment when OOS can arise.
when your ship arrives in europe, the european docks window immediately opens. however, various actions can be performed at the same time - a native raid on the colony (money can be stolen from the colony during the raid), an automated scout can visit the native village (the leader can donate money) or find money when searching ruins or burial grounds. however, while the euro window is open, the changes in the treasury are not applied (at least not visible). and after entering the main map, the amount of money in the treasury changes.

**Comment by raystuttgart** [comment]
But that is just display in the Screen because it may not update constantly - because that would ruin performance.
That does not mean however that the variable in the code is not already updated.

The money variable internally in the DLL should always already be correct.
"Display Issues" cannot cause OOS.

**Summary:**
I really do not believe that "Display Issues" in any way are related to OOS.
The DLL logic does not really care at all what the Python logic displays - it does not even know it.

**Comment by Nightinggale** [comment]
With all the OOS fixes recently I will assume this is fixed by now. If not then it would be better with new savegames.
